File: CopyPaste.Core/UpdateChecker.cs
using System.Net.Http;
using System.Text.Json;

namespace CopyPaste.Core;

/// <summary>
/// Event args for update available notification.
/// </summary>
[System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1054:URI parameters should not be strings",
    Justification = "URL comes from JSON API response and is consumed as string by Process.Start")]
public sealed class UpdateAvailableEventArgs(string newVersion, string downloadUrl) : EventArgs
{
    public string NewVersion { get; } = newVersion;

    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1056:URI properties should not be strings",
        Justification = "URL comes from JSON API response and is consumed as string by Process.Start")]
    public string DownloadUrl { get; } = downloadUrl;
}

/// <summary>
/// Checks for updates by comparing current version against the latest GitHub release.
/// Only runs for standalone (unpackaged) builds — Store builds use Store updates.
/// </summary>
public sealed class UpdateChecker : IDisposable
{
    private const string _gitHubReleasesUrl = "https://api.github.com/repos/rgdevment/CopyPaste/releases/latest";
    private static readonly Uri _gitHubReleasesUri = new(_gitHubReleasesUrl);
    private const string _dismissedVersionFileName = "dismissed_update.txt";

    private static readonly TimeSpan _checkInterval = TimeSpan.FromHours(12);
    private static readonly TimeSpan _httpTimeout = TimeSpan.FromSeconds(15);

    private readonly HttpClient _httpClient;
    private readonly Timer _timer;
    private bool _isDisposed;

    /// <summary>
    /// Raised when a new version is available.
    /// </summary>
    public event EventHandler<UpdateAvailableEventArgs>? OnUpdateAvailable;

    public UpdateChecker()
    {
        _httpClient = new HttpClient { Timeout = _httpTimeout };
        _httpClient.DefaultRequestHeaders.Add("User-Agent", "CopyPaste-UpdateChecker");
        _httpClient.DefaultRequestHeaders.Add("Accept", "application/vnd.github.v3+json");

        _timer = new Timer(
            _ => CheckForUpdateAsync().ConfigureAwait(false),
            null,
            TimeSpan.FromSeconds(30), // First check 30s after startup
            _checkInterval
        );
    }

    /// <summary>
    /// Gets the current app version from the assembly.
    /// </summary>
    public static string GetCurrentVersion()
    {
        var assembly = System.Reflection.Assembly.GetEntryAssembly();
        var informational = assembly?
            .GetCustomAttributes(typeof(System.Reflection.AssemblyInformationalVersionAttribute), false)
            .OfType<System.Reflection.AssemblyInformationalVersionAttribute>()
            .FirstOrDefault()?.InformationalVersion;

        if (!string.IsNullOrEmpty(informational))
        {
            // Remove metadata after '+' (e.g., "1.2.0+abc123" -> "1.2.0")
            var plusIndex = informational.IndexOf('+', StringComparison.Ordinal);
            return plusIndex > 0 ? informational[..plusIndex] : informational;
        }

        var version = assembly?.GetName().Version;
        return version != null ? $"{version.Major}.{version.Minor}.{version.Build}" : "0.0.0";
    }

    internal async Task CheckForUpdateAsync()
    {
        if (_isDisposed) return;

        try
        {
            using var response = await _httpClient.GetAsync(_gitHubReleasesUri).ConfigureAwait(false);

            if (!response.IsSuccessStatusCode)
            {
                AppLogger.Warn($"Update check failed: HTTP {response.StatusCode}");
                return;
            }

            var json = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
            var release = JsonSerializer.Deserialize(json, GitHubReleaseJsonContext.Default.GitHubRelease);

            if (release == null || string.IsNullOrEmpty(release.TagName))
            {
                AppLogger.Warn("Update check: invalid response");
                return;
            }

            var latestVersion = release.TagName.TrimStart('v', 'V');
            var currentVersion = GetCurrentVersion();

            if (IsNewerVersion(latestVersion, currentVersion))
            {
                if (IsVersionDismissed(latestVersion))
                {
                    AppLogger.Info($"Update {latestVersion} available but dismissed by user");
                    return;
                }

                var downloadUrl = release.HtmlUrl ?? $"https://github.com/rgdevment/CopyPaste/releases/tag/{release.TagName}";
                AppLogger.Info($"Update available: {currentVersion} → {latestVersion}");
                OnUpdateAvailable?.Invoke(this, new UpdateAvailableEventArgs(latestVersion, downloadUrl));
            }
            else
            {
                AppLogger.Info($"App is up to date ({currentVersion})");
            }
        }
        catch (HttpRequestException ex)
        {
            AppLogger.Warn($"Update check failed (network): {ex.Message}");
        }
        catch (TaskCanceledException)
        {
            AppLogger.Warn("Update check timed out");
        }
        catch (JsonException ex)
        {
            AppLogger.Warn($"Update check failed (parse): {ex.Message}");
        }
    }

    /// <summary>
    /// Compares two semver-style versions. Returns true if <paramref name="latest"/> is newer.
    /// Supports pre-release tags (e.g., "1.0.0-beta.1" &lt; "1.0.0").
    /// </summary>
    internal static bool IsNewerVersion(string latest, string current)
    {
        // Split off pre-release suffix
        var (latestBase, latestPre) = SplitVersion(latest);
        var (currentBase, currentPre) = SplitVersion(current);

        if (!Version.TryParse(NormalizeVersion(latestBase), out var latestVer) ||
            !Version.TryParse(NormalizeVersion(currentBase), out var currentVer))
        {
            return false;
        }

        var baseComparison = latestVer.CompareTo(currentVer);
        if (baseComparison != 0)
            return baseComparison > 0;

        // Same base version: stable > pre-release
        if (string.IsNullOrEmpty(latestPre) && !string.IsNullOrEmpty(currentPre))
            return true; // latest is stable, current is pre-release

        if (!string.IsNullOrEmpty(latestPre) && string.IsNullOrEmpty(currentPre))
            return false; // latest is pre-release, current is stable

        // Both pre-release: compare numerically (beta.2 > beta.1)
        if (!string.IsNullOrEmpty(latestPre) && !string.IsNullOrEmpty(currentPre))
            return string.Compare(latestPre, currentPre, StringComparison.OrdinalIgnoreCase) > 0;

        return false;
    }

    private static (string baseVersion, string preRelease) SplitVersion(string version)
    {
        var dashIndex = version.IndexOf('-', StringComparison.Ordinal);
        return dashIndex > 0
            ? (version[..dashIndex], version[(dashIndex + 1)..])
            : (version, string.Empty);
    }

    private static string NormalizeVersion(string version)
    {
        var parts = version.Split('.');
        return parts.Length switch
        {
            1 => $"{parts[0]}.0.0",
            2 => $"{parts[0]}.{parts[1]}.0",
            _ => version
        };
    }

    /// <summary>
    /// Persists the user's decision to dismiss a specific version update.
    /// </summary>
    public static void DismissVersion(string version)
    {
        try
        {
            var filePath = GetDismissedFilePath();
            File.WriteAllText(filePath, version);
            AppLogger.Info($"User dismissed update notification for version {version}");
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            AppLogger.Warn($"Failed to save dismissed version: {ex.Message}");
        }
    }

    /// <summary>
    /// Checks if the user has dismissed a specific version notification.
    /// </summary>
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1031:Do not catch general exception types",
        Justification = "Dismissed version check is non-critical - any failure should return false")]
    private static bool IsVersionDismissed(string version)
    {
        try
        {
            var filePath = GetDismissedFilePath();
            if (!File.Exists(filePath)) return false;

            var dismissed = File.ReadAllText(filePath).Trim();
            return string.Equals(dismissed, version, StringComparison.OrdinalIgnoreCase);
        }
        catch
        {
            return false;
        }
    }

    private static string GetDismissedFilePath() =>
        Path.Combine(StorageConfig.ConfigPath, _dismissedVersionFileName);

    public void Dispose()
    {
        if (_isDisposed) return;
        _isDisposed = true;
        _timer.Dispose();
        _httpClient.Dispose();
    }
}

/// <summary>
/// Minimal model for GitHub release API response.
/// </summary>
public sealed class GitHubRelease
{
    [System.Text.Json.Serialization.JsonPropertyName("tag_name")]
    public string? TagName { get; set; }

    [System.Text.Json.Serialization.JsonPropertyName("html_url")]
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1056:URI properties should not be strings",
        Justification = "Mapped from GitHub API JSON response")]
    public string? HtmlUrl { get; set; }

    [System.Text.Json.Serialization.JsonPropertyName("prerelease")]
    public bool Prerelease { get; set; }
}

/// <summary>
/// JSON serialization context for GitHub release (required for trimming).
/// </summary>
[System.Text.Json.Serialization.JsonSerializable(typeof(GitHubRelease))]
[System.Text.Json.Serialization.JsonSourceGenerationOptions(PropertyNameCaseInsensitive = true)]
public partial class GitHubReleaseJsonContext : System.Text.Json.Serialization.JsonSerializerContext
{
}

File: v2/core/lib/services/update_checker.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;

class UpdateInfo {
  const UpdateInfo({
    required this.version,
    required this.downloadUrl,
  });

  final String version;
  final String downloadUrl;
}

class UpdateChecker {
  UpdateChecker({this.configPath});

  static const String _releasesUrl =
      'https://api.github.com/repos/rgdevment/CopyPaste/releases/latest';
  static const Duration _startupDelay = Duration(minutes: 1);
  static const Duration _httpTimeout = Duration(seconds: 15);
  static const String _dismissedFileName = 'dismissed_update.txt';

  final String? configPath;

  final _onUpdateAvailable =
      StreamController<UpdateInfo>.broadcast();

  Stream<UpdateInfo> get onUpdateAvailable =>
      _onUpdateAvailable.stream;

  Timer? _startTimer;
  bool _disposed = false;

  void start(String currentVersion) {
    _startTimer = Timer(_startupDelay, () => _check(currentVersion));
  }

  Future<void> _check(String currentVersion) async {
    if (_disposed) return;
    try {
      final response = await http
          .get(
            Uri.parse(_releasesUrl),
            headers: {
              'User-Agent': 'CopyPaste-UpdateChecker',
              'Accept': 'application/vnd.github.v3+json',
            },
          )
          .timeout(_httpTimeout);

      if (response.statusCode != 200) return;

      final json =
          jsonDecode(response.body) as Map<String, dynamic>;
      final tagName = json['tag_name'] as String?;
      final downloadUrl = json['html_url'] as String?;
      final prerelease = json['prerelease'] as bool? ?? false;

      if (tagName == null || downloadUrl == null) return;
      if (prerelease) return;

      final latest =
          tagName.startsWith('v') ? tagName.substring(1) : tagName;

      if (!_disposed && isNewer(latest, currentVersion)) {
        if (!isVersionDismissed(latest)) {
          _onUpdateAvailable.add(
            UpdateInfo(version: latest, downloadUrl: downloadUrl),
          );
        }
      }
    } catch (_) {}
  }

  void dismissVersion(String version) {
    if (configPath == null) return;
    try {
      final file = File(p.join(configPath!, _dismissedFileName));
      file.parent.createSync(recursive: true);
      file.writeAsStringSync(version);
    } catch (_) {}
  }

  bool isVersionDismissed(String version) {
    if (configPath == null) return false;
    try {
      final file = File(p.join(configPath!, _dismissedFileName));
      if (!file.existsSync()) return false;
      return file.readAsStringSync().trim() == version;
    } catch (_) {
      return false;
    }
  }

  static bool isNewer(String latest, String current) {
    try {
      final l = _parseVersion(latest);
      final c = _parseVersion(current);
      for (var i = 0; i < 3; i++) {
        final lv = i < l.length ? l[i] : 0;
        final cv = i < c.length ? c[i] : 0;
        if (lv > cv) return true;
        if (lv < cv) return false;
      }
      return false;
    } catch (_) {
      return false;
    }
  }

  static List<int> _parseVersion(String version) {
    // Strip pre-release suffix: "1.0.0-beta.1" → "1.0.0"
    final base = version.split('-').first;
    return base.split('.').map(int.parse).toList();
  }

  void dispose() {
    _disposed = true;
    _startTimer?.cancel();
    _onUpdateAvailable.close();
  }
}
